[Infra UI] Add a source settings ui

weltenwort opened the pull request:
## Summary

This adds a configuration UI to change various properties of an infrastructure data source. The properties that can be currently changed are:

* name
* index pattern for metrics
* index pattern for logs
* fields like the timestamp, container/host/pod identification

It is meant as a basis for providing configurability for other source properties like outgoing links and formatting rules in the future.

part of elastic/kibana#26539 

## Screenshots

### Unsaved changes
![image](https://user-images.githubusercontent.com/973741/52202904-17aa5d80-2870-11e9-89e9-1c7603d204ce.png)

### Loading
![image](https://user-images.githubusercontent.com/973741/52202956-3b6da380-2870-11e9-9423-2b10e24c8a87.png)

### Empty Required Field
![image](https://user-images.githubusercontent.com/973741/52202928-285ad380-2870-11e9-9a97-63c829c225a2.png)

### No indices prompt

![image](https://user-images.githubusercontent.com/973741/51904623-69aa3980-23bf-11e9-948b-d75d4ce98f66.png)

### Triggers

#### Waffle map
![image](https://user-images.githubusercontent.com/973741/51702879-3f4e2a00-2015-11e9-9c6b-abac1f126636.png)

#### Logs
![image](https://user-images.githubusercontent.com/973741/52203010-60faad00-2870-11e9-82a0-efb3a78a9c6d.png)

### Checklist

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- ~~This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

weltenwort committed:
Add a source configuration flyout

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2712/)

makwarth commented:
@weltenwort I was expecting the configuration to be on its own page, as opposed to fly-out. Primarily because it's sort of a "danger zone" and because I don't expect users to need to edit the configuration often. Thoughts?

weltenwort commented:
It can be its own page, the implementation does not depend on it being a flyout. The main difference would be wether its easy to return to the original page that prompted the desire to change the settings.

It might accelerate the workflow of "see something is wrong" -> "tweak the settings" -> "check whether it improves the situation" -> "repeat" cycle to have it as an "overlay".

I can see that happen relatively often if something like the links and the log formatting are configured there.

makwarth commented:
It seems like we should  have some configuration options easy accessible in a fly-out (log format, wrap lines, font size), and some should be on a dedicated configuration page (index patterns).

weltenwort commented:
Would that mean that we also start persisting the settings that are transient at the moment (like the wrapping and font size)? Mixing two persistence paradigms in one settings flyout could be confusing.

makwarth commented:
Good question. I don't have all of the context, so I'm curious to why font size and wrapping shouldn't persist. I would imagine people adjusting font size once/occasionally and then sticking to it?

weltenwort commented:
I see that as something that is specific to the user's screen and window management setup, so I would expect these settings to differ between users.

makwarth commented:
Right, don't know why I just thought of them being configured on the user level. Need coffee!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5481/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5527/)

legrego reviewed (commented):
The changes that touch spaces/security LGTM -- don't feel like you need a formal review from us unless you change something else in those plugins, or if you want our input on anything specific

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7744/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8217/)

fbaligand commented:
Hi,
It seems that in the settings UI, we can not customize “message” field.
Is it possible to add it ?

weltenwort commented:
> It seems that in the settings UI, we can not customize “message” field. Is it possible to add it ?

Given the complexity of the log formatting, that won't be possible in this iteration. The log message is actually extracted from the document using a heuristic that can not be expressed with a single setting. If I get my way, adding a UI for that is my next task, though. :wink:

fbaligand commented:
Nice to see it is your next task!
BTW, I'm very surprised to learn that "log message" in Logs UI is not just "message" field but is computed using heuristic.
Can you explain ?

weltenwort commented:
The Logs UI attempts to construct the messages from various fields in the documents by applying a series of rules (currently https://github.com/elastic/kibana/tree/b0e08dce8e64b7fd26527bda029ff59f3c7401e5/x-pack/plugins/infra/server/lib/domains/log_entries_domain/builtin_rules) to the documents. This is done in order to...

* support and normalize diverse input formats (e.g. apache vs nginx access logs)
* benefit from parsing that has happened during ingestion (e.g. by filebeat or logstash)
* make the displayed message configurable via the UI (which is what I'm hoping to implement after this PR is done)

fbaligand commented:
OK!
Thanks for the explanation!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8408/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8877/)

weltenwort commented:
looks like a jenkins problem - jenkins, test this again

```
00:00:00.056 FATAL: java.io.InvalidClassException: hudson.util.Secret; local class incompatible: stream classdesc serialVersionUID = 1, local class serialVersionUID = 6967577727317184689
00:00:00.057 Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from kibana-ci-immutable-centos-1548429376126207066.c.elastic-ci-prod.internal/10.224.1.12:46772
00:00:00.057 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1741)
00:00:00.057 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
00:00:00.057 		at hudson.remoting.Channel.call(Channel.java:955)
00:00:00.057 		at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:283)
00:00:00.057 		at com.sun.proxy.$Proxy107.addCredentials(Unknown Source)
00:00:00.057 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl.addCredentials(RemoteGitImpl.java:200)
00:00:00.057 		at hudson.plugins.git.GitSCM.createClient(GitSCM.java:842)
00:00:00.057 		at hudson.plugins.git.GitSCM.createClient(GitSCM.java:812)
00:00:00.057 		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1180)
00:00:00.057 		at hudson.scm.SCM.checkout(SCM.java:504)
00:00:00.057 		at hudson.model.AbstractProject.checkout(AbstractProject.java:1208)
00:00:00.057 		at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:574)
00:00:00.057 		at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
00:00:00.057 		at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:499)
00:00:00.057 		at hudson.model.Run.execute(Run.java:1810)
00:00:00.057 		at hudson.matrix.MatrixRun.run(MatrixRun.java:153)
00:00:00.057 		at hudson.model.ResourceController.execute(ResourceController.java:97)
00:00:00.057 		at hudson.model.Executor.run(Executor.java:429)
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9044/)

weltenwort committed:
Remove client-side ts-optchain usage

makwarth commented:
@weltenwort this looks great! Few comments:
- Logs UI: Can we move the UI element to the right of the search bar, so we have `[source config] -[customize]`
- Can we rename "Unnamed Source" to "Default source" (or maybe just "Default")
  - I'm not sure how to reset, so I only tried this once: It seemed like before renaming the source the flyout button said "Create new.., and then switched to "Update..". Users can only have one source atm, right? If yes, the button could just say "Update source" all the time?
- Changing "@timestamp" doesn't seem to have an impact?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9397/)

weltenwort committed:
Move source config button in logs toolbar

weltenwort committed:
Change default source name to "Default"

weltenwort committed:
Remove distinction between source creation and update

weltenwort commented:
Thanks for the feedback, @makwarth.

> Logs UI: Can we move the UI element to the right of the search bar, so we have `[source config] -[customize]`

done

![image](https://user-images.githubusercontent.com/973741/51903768-4aaaa800-23bd-11e9-9417-de517e79aa29.png)

I feel we should reorganize the toolbar soon to find better room for these buttons. Maybe a two-row layout as you suggested in #27916 would allow us to break the current "Customize" menu into separate toolbar buttons for scale, text wrap and font size on the second row. :thinking:

> Can we rename "Unnamed Source" to "Default source" (or maybe just "Default")

also done (see screenshot above)

> I'm not sure how to reset, so I only tried this once: It seemed like before renaming the source the flyout button said "Create new.., and then switched to "Update..". Users can only have one source atm, right? If yes, the button could just say "Update source" all the time?

Right, while technically the first change is a "create" operation, this distinction is not meaningful to the user at the moment. I unified both cases visually into an "update".

> 
> Changing "@timestamp" doesn't seem to have an impact?

It does, but the field changes do not automatically trigger a reload at the moment. I guess we would want it to though?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9452/)

makwarth commented:
Agree, we definitely need to redesign the toolbar. 

>I unified both cases visually into an "update".

Great

>It does, but the field changes do not automatically trigger a reload at the moment. I guess we would want it to though?

Yes, or at least that seems to be the case when changing the index name?

weltenwort committed:
Add unit test for source configuration updater

weltenwort committed:
Remove derelict EuiDescribedFormGroup type definition

weltenwort committed:
Include new default source name in api test

weltenwort commented:
I would be much obliged if any of the reviewers with a working screen-reader setup could test that aspect of the source configuration flyout for me. I failed to come up with a working setup to test this on. :see_no_evil:

Kerry350 commented:
> I would be much obliged if any of the reviewers with a working screen-reader setup could test that aspect of the source configuration flyout for me. I failed to come up with a working setup to test this on.

I have a working ChromeVox extension. I can try this out later 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9716/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9722/)

Kerry350 commented:
Screenreader wise the flyout works great (with ChromeVox anyway) 👍

weltenwort commented:
Thank you for checking!

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9837/)

skh commented on x-pack/plugins/infra/public/components/source_configuration/source_configuration_flyout.tsx:
```diff
@@ -0,0 +1,168 @@
+          {({ create, configuration, exists, isLoading, update }) =>
+            configuration ? (
+              <WithSourceConfigurationFormState
+                initialState={{
```
I'm wondering if it is better to let the container handle its `initialState` internally, instead of passing it in.
Do you think we will use `WithSourceConfigurationFormState` with different initial states, ever?

skh replied:
Result of the offline discussion: I misunderstood what `initialState` contains here, it is the state of the form before the user made any changes, and not the fallback to the initial / default values of a source configuration.

A less generic name would maybe be helpful, if you can find a good one.

weltenwort replied:
Maybe `initialFormState` clarifies that the "initial" only refers to the state of the form and not the default configuration?

skh replied:
Yes, thank you!

skh reviewed (changes requested):
I found two issues during testing:

* when I click into any input field and use `backspace` (or triple-click and then `backspace`) to empty the field before entering a new value, the original value is filled in immediately when the field should stay empty. I found this a bit annoying.
* after changing a field and saving the updated configuration, the currently open page does not reload its data, even though it is very likely that the changes in the source configuration affect what's displayed. The reload happens as expected when I change the `logAlias` or `metricAlias`.

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

weltenwort committed:
Make form state property name more specific

weltenwort committed:
Properly handle empty strings in source update operations

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10099/)

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

weltenwort committed:
Remove hardcoded source id from options container

weltenwort committed:
Reload page content after reloading source configuration

weltenwort commented:
Very good catches, @skh. Thank you!

I hope to have addressed your points from above with my latest set of commits:

* renamed the property to `initialFormState`
* treat empty strings as valid values for the inputs (including revamped unit tests)
* prioritize the source loading indicators in order to force refetch after source configuration changes

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10196/)

simianhacker reviewed (changes requested):
I think we should at least make the name a required field on the form. Right now you can delete all the fields on the form and submit it. This leads to the name being empty. Luckily you can still click on the cog icon to open it. I wonder if maybe we should explain with a sentence or use the note option on the fields to let the user know what the defaults are if those fields are empty, logging seems to work just fine. Whoops, It looks like if you delete the fields values (timestamp, host, container, docker) it breaks the waffle map and throws an error in the console.

weltenwort commented:
The non-empty constraint makes sense on every field, I guess. I'll change it to disable the save button when any field is empty.

simianhacker commented:
After thinking about it a little I think it would be helpful to add `helpText` attributes to each field that has a description similar to what we have in the docs. I would use https://www.elastic.co/guide/en/kibana/current/infrastructure-ui-settings-kb.html

weltenwort committed:
Add empty string validation to source config form

weltenwort committed:
Add help text to source configuration fields

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

weltenwort commented:
@skh @simianhacker I updated the PR with

* validation to prevent saving with an empty field
* help texts for the inputs (except for "name", which seemed kinda obvious)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10385/)

simianhacker commented:
LGTM... GOOD JOB!

skh reviewed (approved):
This is much nicer now, thank you :+1: 

One little nitpick I have left (applies to all fields):

![image](https://user-images.githubusercontent.com/189073/52198654-d2802e80-2863-11e9-88c6-d638a06f8326.png)

If there is a default value mentioned, I'd assume that I can leave the field empty and it would fall back to the default when I do, but that clashes with the warning that the field shouldn't be empty. Can this be worded differently?

weltenwort committed:
Change help text wording from default to recommended

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

weltenwort committed:
Merge branch 'master' into infra-ui-enhancement-source-configuration-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10626/)
